### PR TITLE
Improve model evaluation and object proposals generation

### PR DESCRIPTION
**Pull request by nagitsu**
- Fixes Faster R-CNN object proposals generation by not artificially limiting predictions.
- Fix mAP calculation code; now limiting by `max_detections` instead of `min_probability`.
- Implements COCO metric instead of Pascal VOC's, calculating mAP at different IoU thresholds (`[0.50:0.05:0.95]`).

### CODE REVIEW

File: luminoth/eval.py [hunk 1]
```diff
@@ -53,19 +52,20 @@ def eval(dataset_split, config_files, watch, from_global_step,
     # Disable data augmentation.
     config.dataset.data_augmentation = []
 
+    # Override max detections with specified value.
+    config.model.rcnn.proposals.total_max_detections = max_detections
```
Comment (vierja) [inline luminoth/eval.py]: Does this break support por RPN only? Maybe add `post_nms_top_n` (renaming it while at it) support?

File: luminoth/eval.py [hunk 2]
```diff
@@ -18,10 +18,9 @@
 @click.option('--from-global-step', type=int, default=None, help='Consider only checkpoints after this global step')  # noqa
 @click.option('override_params', '--override', '-o', multiple=True, help='Override model config params.')  # noqa
 @click.option('--files-per-class', type=int, default=10, help='How many files per class display in every epoch.')  # noqa
-@click.option('--iou-threshold', type=float, default=0.5, help='IoU threshold to use.')  # noqa
-@click.option('--min-probability', type=float, default=0.5, help='Minimum probability to use.')  # noqa
```
Comment (vierja) [inline luminoth/eval.py]: Is there no use case or metric that actually uses this thresholding for evaluation?

File: luminoth/eval.py [hunk 3]
```diff
@@ -376,42 +365,71 @@ def evaluate_once(config, writer, saver, ops, checkpoint,
...
+                )
+            )
+            tf.logging.info(
+                'Average Recall (AR) @ [0.50:0.95] = {:.3f}'.format(
+                    mar_at_range
+                )
+            )
+
+            for idx, val in enumerate(ap_per_class[:, 0]):
+                tf.logging.debug(
```
Comment (vierja) [inline luminoth/eval.py]: Do some of your Python formatting magic for doing something similar to:
image
?

File: luminoth/eval.py (hunk 1 above)
Comment (nagitsu) [reply luminoth/eval.py]: Noted.

File: luminoth/eval.py (hunk 2 above)
Comment (nagitsu) [reply luminoth/eval.py]: Not really, `min_probability` should never be limited.

File: luminoth/eval.py [hunk 4]
```diff
@@ -492,23 +490,23 @@ def calculate_map(output_per_batch, num_classes, iou_threshold=0.5):
...
-        The "difficult example" flag of VOC dataset is being ignored.
-
-    Todo:
-        * Use VOC2012-style for integrating the curve. That is, use all recall
-          points instead of a fixed number of points like in VOC2007.
+        (``np.ndarray``, ``ndarray``) tuple. The first value is an array of
+        size (`num_classes`,), with the AP value per class, while the second
+        one is an array for the AR.
 
     .. _VOC mAP metric:
```
[the remark this answers is not in the document]
Comment (nagitsu) [reply luminoth/eval.py]: That link is still the most comprehensive available. COCO doesn't seem to have a detailed explanation of the calculations done.

**Pull request merged**